### ISSUE TITLE
probably false positives with navigation.setOptions inside useLayoutEffect on react-native

### ISSUE THREAD

**Issue by efstathiosntonas** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#10]
I get this false positive imo:

```tsx
  useLayoutEffect(() => {
    navigation.setOptions({
      headerLeft: () => <CloseButtonHeader disabled={submitting} onPress={onBack} />,
      headerRight: () => (
        <NewPostSaveButton
          disabled={submitting || loading}
          isEdit={!!post}
          onPress={onPressSave}
          title={
            post
              ? t({
                  message: 'Save',
                  context: 'action'
                })
              : t({
                  message: 'Post',
                  context: 'action'
                })
          }
        />
      )
    });
  }, [loading, navigation, onBack, onPressSave, post, submitting, t]);
```

throws: 

```
ESLint: Avoid coupling parent behavior or state to a child component. Instead, lift shared logic or state up to the parent. (react-you-might-not-need-an-effect/you-might-not-need-an-effect)
```

It's documented in here: https://reactnavigation.org/docs/navigation-object#setoptions

**Comment by efstathiosntonas** [comment]
well, the react-navigation v6.x has it like my code: https://reactnavigation.org/docs/6.x/navigation-prop#setoptions


```tsx
 function ProfileScreen({ navigation, route }) {
  const [value, onChangeText] = React.useState(route.params.title);

  React.useEffect(() => {
    navigation.setOptions({
      title: value === '' ? 'No title' : value,
    });
  }, [navigation, value]);

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <TextInput
        style={{ height: 40, borderColor: 'gray', borderWidth: 1 }}
        onChangeText={onChangeText}
        value={value}
      />
      <Button title="Go back" onPress={() => navigation.goBack()} />
    </View>
  );
}
```

**Comment by NickvanDyke** [comment]
No no I think you're on to something!

It did occur to me that this could be a library HOC issue. In that case, the developer doesn't have enough control to lift state/logic to the parent, because the parent is a library. So the HOC prop is technically "external" state, which I can update the plugin to handle when it sees the prop wrapped in a HOC function.

Unfortunately in your particular example, I don't see any indication in the source code that `ProfileScreen` uses a HOC. It seems like it's just an implied trait of React Native that it can receive `navigation` because it's named `xScreen` and presumably passed to some React Native render function at the application root? So I'm not sure how we could know to consider `navigation` as external here 🤔

**Comment by efstathiosntonas** [comment]
On react-native this: 

```
 <IndexStack.Screen
          getComponent={() => PostDetail}
          name="PostDetail"
          options={{
            headerBackButtonDisplayMode: 'minimal'
          }}
        />
```

will pass the `navigation` and `route` prop to the component

**Comment by NickvanDyke** [comment]
Thanks @efstathiosntonas, that makes sense. I'll think about a good solution for this. Ultimately if the plugin continues to grow in popularity, I may just add a configuration mechanism for known edge-cases like this. I appreciate you calling it out, as well as the broader HOC issue.

In the meantime, and maybe you've already done this, but you could skip running this plugin on these `Screen` files if they have a consistent directory or naming via [ESLint ignores](https://eslint.org/docs/latest/use/configure/ignore)!

**Comment by NickvanDyke** [comment]
@efstathiosntonas I just updated the plugin to skip `useLayoutEffect`s after seeing https://github.com/NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect/issues/25. Does that fix this issue too? i.e. is React Native's `navigation` always supposed to be called in `useLayoutEffect`, not `useEffect`?

**Comment by efstathiosntonas** [comment]
yes!! updated earlier and eslint warnings where reduced by 120 lol! 

Thanks Nick!
